pkg/git/git.go

```diff
@@ -0,0 +1,69 @@
+// Copyright The Shipwright Contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package git
+
+import (
+	"fmt"
+	"github.com/go-git/go-git/v5/config"
+	"github.com/go-git/go-git/v5/plumbing/transport"
+	"github.com/go-git/go-git/v5/storage/memory"
+	"go.uber.org/zap"
+	"regexp"
+
+	gitpackage "github.com/go-git/go-git/v5"
+)
+
+const defaultRemote = "origin"
+
+var (
+	// UrlRegexFormat matches the url of git repository
+	httpUrlRegexFormat        = regexp.MustCompile(`^((http:\/\/)|(https:\/\/))?([a-zA-Z0-9]([a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])?\.)+[a-zA-Z]{2,6}(\/)`)
+	sshUrlRegexFormat         = regexp.MustCompile(`(ssh:\/\/[\w\d\.]+|.+@?.+\..+:)(:[\d]+){0,1}\/*(.*)`)
+	httpsPublicUrlRegexFormat = regexp.MustCompile(`(https:\/\/)?(github\.)+com(\/)`)
+	sshPublicUrlRegexFormat   = regexp.MustCompile(`(git+@github.com:)(:[\d]+){0,1}/*(.*)`)
+)
+
+func ValidateGitURLExists(urlPath string) error {
+	repoKind := ValidateGitURLFormat(urlPath)
+	prod, _ := zap.NewProduction()
+	logger := prod.Sugar()
+	if repoKind == "public" {
+		repo := gitpackage.NewRemote(memory.NewStorage(), &config.RemoteConfig{
+			Name: defaultRemote,
+			URLs: []string{urlPath},
+		})
+		// Note: When the urlPath is an valid public path, however, this path doesn't exist, func will return `authentication required`, this maybe mislead
+		// So convert this error message to `repository not found`
+		_, err := repo.List(&gitpackage.ListOptions{})
+		if err != nil && err == transport.ErrAuthenticationRequired {
```

Comment:
@qu1queee Indeed, they have clear error definitions. But as I said in [here](https://github.com/shipwright-io/build/pull/512#discussion_r539975582). If a valid and public repo doesn't exist, `List` func will return `authentication required` error rather than `repository not found` error. Only you try to verify an invalid URL, like `foobar`, then `List` func will return `repository not found`.

That means now, for a sourceURL has a valid format, if `List` func doesn't find it in remote, then it will think the authentication is incorrect or you don't set authentication.

Another way to avoid this is to use [`Clone`](https://github.com/go-git/go-git/blob/master/repository.go#L197) func to verify sourceURL, but I am wondering this method will make our controller slow, because it needs to clone the whole repo.

Not sure this makes sense or not?

Reply:
yes I think it make sense, but let me see this, testing locally atm.

Reply:
@xiujuan95 ok now it make sense to me. The `authentication required` is because when writing something that does not exist, e.g. `https://github.com/shipwright-io/buildfoobar`, I assume git thinks is a private repo, and tries to authenticate or similar. I did a comparison with a plain `git ls-remote https://github.com/shipwright-io/buildfoobar` and git ask for user/pwd. So its fine to do this. As you said, a single error msg, for both `authentication required` and `repository not found`.